docs/async/async-in-depth.md

```diff
@@ -0,0 +1,151 @@
+
+For example, in Windows an OS thread makes a call to the network device driver and asks it to perform the networking operation via an Interrupt Request Packet (IRP) which represents the operation.  The device driver recieves the IRP, makes the call to the network, marks the IRP as "pending", and returns back to the OS.  Because the OS thread now knows that the IRP is "pending", it doesn't have any more work to do for this job and "returns" back so that it can be used to perform other work.
+
+![](../../images/async/io-1.jpg)
+(caption) A view into what happens when an async call is made.
+
+When the request is fulfilled and data comes back through the device driver, it notifies the CPU of new data received via an interrupt.  How this interrupt gets handled will vary depending on the OS, but eventually the data will be passed through the OS until it reaches a system interop call (for example, in Linux an interrupt handler will schedule the bottom half of the IRQ to pass the data up through the OS asynchronously).  Note that this *also* happens asynchronously!  The result is queued up until the next available thread is able execute the async method and "unwrap" the result of the completed task.
+
+![](../../images/async/io-2.jpg)
+(caption) A view into what happens when an I/O request is complete.
+
+Throughout this entire process, a key takeaway is that **no thread is dedicated to running the task**.  Tasks have no thread affinity.  Although work is executed in some contexts (i.e. the OS does have to pass data to a device driver and respond to an interrupt), there is no thread dedicated to *waiting* for data from the request to come back.  This allows the system to handle a much larger volume of work rather than waiting for some I/O call to finish.
+
+Although the above may seem like a lot of work to be done, when measured in terms of wall clock time, it’s miniscule compared to the time it takes to do the actual I/O work. Although not at all precise, a potential timeline for such a call would look like this:
+
+0-1————————————————————————————————————————————————–2-3
+
+*   Time spent from points `0` to `1` is everything up until an async method yields control to its caller.
+*   Time spent from points `1` to `2` is the time spent on I/O, with no CPU cost.
+*   Finally, time spent from points `2` to `3` is passing control back (and potentially a value) to the async method, at which point it is executing again.
+
+### What does this mean for a server scenario?
+
+This model works well with a typical server scenario workload.  Because tasks don't have an affinity to threads, it means that the server threadpool can use any thread for any task. As a result, it can service a much higher volume of web requests than if each thread were dedicated to running a particular request.
```

Comment:
Here, it is correct that the server threadpool can use any thread for any task. That's because they are all threadpool threads. In client scenarios, there is a distinction made for the UI thread.